canonical/snapcraft.io
# Wd 15262 improve performance of metric page

## Done

- Added pagination when the weekly active devices metric is fetched and if the requested period is more than 3 months.
If the requested period is less than 3 months data no pagination is applied.
- if there are too much data (in that case more than 500 data (ie more than 500 version exists for the requested page)) data is down-sampled. LTTB algorithm is used to select the most meaningful points that represents the dataset for each page. This approach solved most of the "Bad Gateway" error when for example 2 years of data is tried to be fetched. 
- Separated the active latest devices metric endpoint from the weekly active devices end point to make pagination possible.

## How to QA
- go to https://snapcraft-io-4863.demos.haus/
- open a published snap's metric page (https://snapcraft-io-4863.demos.haus/<published-snap-name>/metrics)
- Verify the "Weekly active devices" metric is shown (chart and number)
- Verify the "Territories" metrics are shown
- Change the period filter to something long such as "5 years". (if the data set for the snap is large it might take some time)
- Verify "Weekly active devices" chart is updated
- Change the type filter 
- Verify  "Weekly active devices" chart is updated
- Open a snap that doesnt have any users (such as misk-test)
- Verify "No data found" is shown for all of the metrics
- Verify the filters are disabled
- Verify an info bar is displayed

## Testing

## Issue / Card
Fixes [#15262](https://warthogs.atlassian.net/browse/WD-15262)

## Screenshots

![Screenshot from 2024-10-03 10-47-59](https://github.com/user-attachments/assets/bc410d47-09e2-4b34-8dd8-68da08a9be6e)
![Screenshot from 2024-10-03 10-48-44](https://github.com/user-attachments/assets/db1fea10-2dfc-4d0f-b237-a6e510478d29)
![Screenshot from 2024-10-03 10-50-11](https://github.com/user-attachments/assets/6079af00-2dce-47c7-91ea-de06b7be79fe)
